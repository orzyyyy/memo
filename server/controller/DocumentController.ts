import { Controller, Request } from '../utils/decorator';
import md5 from 'blueimp-md5';
import {
  getWriteMappingPaths,
  updateSider,
  DocumentCategoryProps,
} from '../utils/document';
import DocumentService from '../service/DocumentService';
import { Context } from 'koa';

export interface MappingProps {
  [x: string]: any;
  createTime?: number;
  modifyTime?: number;
  id: string;
  title?: string;
  url?: string;
  type?: string;
  subType?: string;
  category?: DocumentCategoryProps;
  visible?: boolean;
}

@Controller('/document')
export default class DocumentController {
  constructor(private service: DocumentService) {
    this.service = new DocumentService();
  }

  @Request({ url: '/update', method: 'post' })
  async updateTargetDocument(ctx: Context) {
    const {
      layout,
      id,
      title,
      type,
      subType,
      category,
      format,
    } = ctx.request.body;
    if (!id) {
      throw Error('id is undefined');
    }
    const service = this.service;
    // formatted by prettier
    if (format) {
      service.formattedByPrettier(layout);
    }
    // update mapping
    service.updateMapping({
      id,
      title,
      modifyTime: new Date().getTime(),
      type,
      subType,
      category,
    });
    updateSider();
  }

  @Request({ url: '/add', method: 'post' })
  async initDocument(ctx: Context) {
    const { title, type, subType, category } = ctx.request.body;
    const service = this.service;
    const timeStamp = new Date().getTime();
    const id = md5(timeStamp.toString());
    const writeFilesPaths = getWriteMappingPaths(category, id);
    service.initHtmlTemplate(category, id);
    service.updateMapping({
      id,
      title,
      type,
      subType,
      category,
    });
    service.updateContent(category, writeFilesPaths);
    updateSider();
    return id;
  }

  @Request({ url: '/delete', method: 'delete' })
  async deleteTargetDocument(ctx: Context) {
    const { id, category } = ctx.request.body;
    const service = this.service;
    const writeFilesPaths = getWriteMappingPaths(category, id);
    service.deleteTargetDocument(writeFilesPaths);
    service.updateMapping({ id }, true);
    updateSider();
  }

  @Request({ url: '/hide', method: 'post' })
  async hideTargetDocument(ctx: Context) {
    const { id } = ctx.request.body;
    const service = this.service;
    const item = service.getOriginMappingItem(id);
    item.visible = false;
    service.updateMapping(item);
    updateSider();
  }
}
