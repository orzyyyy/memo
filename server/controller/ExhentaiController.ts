import { Controller, Request } from '../utils/decorator';
import { writeIntoJsonFile, getTimeStamp, readJsonFile, joinWithRootPath } from '../utils/common';
import ExhentaiService from '../service/ExhentaiService';
import {
  getLatestListInfo,
  getLatestListFileName,
  getListFiles,
  getLatestDownloadDirName,
  getMissedImgInfo,
} from '../utils/exhentai';
import { Context } from 'koa';
import { getLogger } from '..';
const logger = getLogger('server/controller/ExhentaiController.ts');

export interface ExHentaiInfoItem {
  name?: string;
  detailUrl: string;
  postTime: number;
  thumbnailUrl: string;
}

@Controller('/exhentai')
export default class ExhentaiController {
  constructor(private service: ExhentaiService) {
    this.service = new ExhentaiService();
  }

  @Request({ url: '/', method: 'get' })
  async getThumbnaiInfo() {
    const service = this.service;
    await service.initBrowser();
    const latestListInfo = await getLatestListInfo();
    const results = await service.fetchListInfo(latestListInfo);
    const createTime = getTimeStamp();
    [`src/assets/exhentai/${createTime}`, `dist/assets/exhentai/${createTime}`].map(item =>
      writeIntoJsonFile(item, results),
    );
    logger.info('fetch completed');
    return `./assets/exhentai/${createTime}.json`;
  }

  @Request({ url: '/getLatestSet', method: 'get' })
  async getLatestExHentaiSet() {
    return `./assets/exhentai/${getLatestListFileName()}.json`;
  }

  @Request({ url: '/download', method: 'post' })
  async downloadImages(ctx: Context) {
    const { url } = ctx.request.body;
    const service = this.service;
    await service.initBrowser();
    await service.gotoTargetPage(url, true);
    const prefixPath = await service.ensureFolderForSave();

    logger.info(`start fetching thumbnai urls from: ${url}`);

    const list: string[] = [url, ...(await service.getUrlFromPaginationInfo())];
    const thumbnailUrls = await service.getThumbnailUrlFromDetailPage(list);
    writeIntoJsonFile(`${prefixPath}/list`, thumbnailUrls);

    logger.info('start fetching target images');

    const images = await service.fetchImageUrls(thumbnailUrls);

    logger.info('fetch all image completed');

    writeIntoJsonFile(`${prefixPath}/detail`, images);

    await service.downloadImages(images, prefixPath);
  }

  @Request({ url: '/download/target', method: 'get' })
  async downloadFromExistUrls(ctx: Context) {
    const { name, dateStamp } = ctx.query;
    const service = this.service;
    const prefixPath = `exhentai/${dateStamp}/${name}`;
    const detail = readJsonFile(joinWithRootPath(`${prefixPath}/detail.json`));
    service.downloadImages(detail, prefixPath);
  }

  @Request({ url: '/download/status', method: 'get' })
  async checkMissedDownloadImgs(ctx: Context) {
    const { dateStamp } = ctx.query;
    const latestDirName = getLatestDownloadDirName(dateStamp);
    const latestDirPath = joinWithRootPath(`exhentai/${latestDirName}`);
    const result = getMissedImgInfo(latestDirPath);
    return result;
  }

  @Request({ url: '/dateSet', method: 'get' })
  async getDateSet() {
    const listFiles = getListFiles().reverse();
    const result: { key: string; name: string }[] = [];

    for (const item of listFiles) {
      result.push({ key: `./assets/exhentai/${item}.json`, name: item });
    }

    return listFiles;
  }
}
