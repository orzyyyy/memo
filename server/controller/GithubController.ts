import { Controller, Request } from '../utils/decorator';
import { Context } from 'koa';
import { runCmd } from 'nino-cli/dist/utils/common';
import { joinWithRootPath } from '../utils/common';

@Controller('/github')
export default class MainPageController {
  @Request({ url: '/update', method: 'get' })
  async update(ctx: Context) {
    const { token } = ctx.query;
    if (!token) {
      return 'failed';
    }
    runCmd('sh', [joinWithRootPath('scripts/update.sh')]);
    return 'success';
  }
}
