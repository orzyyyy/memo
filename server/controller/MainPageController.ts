import fs from 'fs-extra';
import path from 'path';
import { Controller, Request } from '../utils/decorator';

@Controller('/')
export default class MainPageController {
  @Request({ url: '/', method: 'get' })
  async getMainPage(ctx: any) {
    ctx.type = 'html';
    ctx.response.body = fs.createReadStream(
      path.join(process.cwd(), 'dist', 'index.html'),
    );
  }
}
