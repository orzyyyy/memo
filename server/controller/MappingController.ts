import fs from 'fs-extra';
import path from 'path';
import { updateMappingRouter } from './MarkdownController';
import { Controller, Request } from '../utils/decorator';

@Controller('/del')
export default class MappingController {
  @Request({ url: '/mapping', method: 'delete' })
  async del(ctx: any) {
    const { id, category } = ctx.request.body;
    const mappingPaths = [
      `src/assets/mapping/${id}.json`,
      `dist/assets/mapping/${id}.json`,
    ];
    const markdownPaths = [
      `src/assets/markdown/${id}.md`,
      `dist/assets/markdown/${id}.md`,
    ];
    let targetPaths: string[] = [];

    switch (category) {
      case 'mapping':
        targetPaths = mappingPaths;
        break;

      case 'markdown':
        targetPaths = markdownPaths;
        break;

      default:
        break;
    }

    try {
      for (const item of targetPaths) {
        if (fs.existsSync(item)) {
          fs.unlinkSync(path.join(process.cwd(), item));
        } else {
          throw Error(`${item} doesn't exist.`);
        }
      }
      updateMappingRouter({ id } as any, true);
      ctx.response.body = true;
    } catch (error) {
      ctx.response.body = false;
    }
  }
}
