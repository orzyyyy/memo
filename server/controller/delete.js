const fs = require('fs');
const path = require('path');

const del = async ctx => {
  const body = ctx.request.body;
  const id = body.id;
  const writeFiles = [`src/layout/${id}.json`, `dist/layout/${id}.json`];
  try {
    for (let item of writeFiles) {
      if (fs.existsSync(item)) {
        fs.unlinkSync(path.join(process.cwd(), item));
      } else {
        // eslint-disable-next-line
        throw Error("file doesn't exist.");
      }
    }
    ctx.response.body = true;
  } catch (error) {
    ctx.response.body = false;
  }
};

module.exports = {
  'DELETE /del': del,
};
