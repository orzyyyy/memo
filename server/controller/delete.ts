import fs from 'fs-extra';
import path from 'path';
import Save from './save';
const { updateMappingRouter } = Save;

const del = async (ctx: any) => {
  const { id, category } = ctx.request.body;
  const mappingPaths = [
    `src/assets/mapping/${id}.json`,
    `dist/assets/mapping/${id}.json`,
  ];
  const markdownPaths = [
    `src/assets/markdown/${id}.md`,
    `dist/assets/markdown/${id}.md`,
  ];
  let targetPaths: string[] = [];

  switch (category) {
    case 'mapping':
      targetPaths = mappingPaths;
      break;

    case 'markdown':
      targetPaths = markdownPaths;
      break;

    default:
      break;
  }

  try {
    for (const item of targetPaths) {
      if (fs.existsSync(item)) {
        fs.unlinkSync(path.join(process.cwd(), item));
      } else {
        throw Error(`${item} doesn't exist.`);
      }
    }
    updateMappingRouter({ id } as any, true);
    ctx.response.body = true;
  } catch (error) {
    ctx.response.body = false;
  }
};

export default { 'DELETE /del/mapping': del };
