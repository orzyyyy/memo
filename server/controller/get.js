const fs = require('fs');
const path = require('path');

const get = async ctx => {
  ctx.type = 'html';
  ctx.response.body = fs.createReadStream(
    path.join(process.cwd(), 'index.html'),
  );
};

module.exports = {
  'GET /': get,
};
