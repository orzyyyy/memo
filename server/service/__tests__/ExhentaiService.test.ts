import Service from '../ExhentaiService';
const {
  goto,
  mockDetail,
  mockImgUrl,
  mockComicNamePrefix,
  mockComicNameSuffix,
} = require('../__mocks__/puppeteer-core');
const { ensureDirSync, ensureFileSync } = require('../__mocks__/fs-extra');
import MockDate from 'mockdate';
import { ExHentaiInfoItem } from '../../controller/ExhentaiController';

jest.mock('../../middleware/DecoratorRouter', () => jest.fn);

describe('ExhentaiService', () => {
  let service: any;

  beforeEach(async () => {
    service = new Service();
    await service.initBrowser();
    MockDate.set(new Date('2019-04-09T00:00:00'));
  });

  afterEach(() => {
    MockDate.reset();
    service = null;
  });

  it('initBrowser and setExHentaiCookie', async () => {
    const result = await service.initBrowser();
    expect(service.page).toEqual(result.page);
  });

  it('gotoTargetPage', async () => {
    await service.gotoTargetPage('test');
    expect(goto).toHaveBeenCalledTimes(1);
    await service.gotoTargetPage('test', true);
    expect(goto).toHaveBeenCalledTimes(3);
  });

  it('getComicName', async () => {
    const result = await service.getComicName();
    expect(result).toBe('test222222222222222222222222222111111111');
    expect(result.length).toBe(40);
  });

  it('ensureFolderForSave', async () => {
    const result = await service.ensureFolderForSave();
    expect(result).toBe('exhentai/20190409/test222222222222222222222222222111111111');
    expect(ensureDirSync).toHaveBeenCalled();
  });

  it('getListInfo', async () => {
    const result = await service.getListInfo(0, {}, '');
    expect(result).toEqual({
      currentResult: mockDetail,
      failed: false,
    });
  });

  it('handleFetchWithFailed', async () => {
    let getData = jest.fn().mockImplementation(() => ({
      currentResult: mockDetail,
      failed: false,
    }));
    let result: any = await service.handleFetchWithFailed(1, {}, 'test', getData);
    expect(result).toEqual(mockDetail);
    getData = jest.fn().mockImplementation(() => ({
      currentResult: mockDetail,
      failed: true,
    }));
    result = await service.handleFetchWithFailed(1, null, 'test', getData);
    expect(result).toBe(false);
    getData = jest.fn().mockImplementation(() => []);
    result = await service.handleFetchWithFailed(1, null, 'test', getData);
    expect(result).toEqual(undefined);
  });

  it('fetchListInfo', async () => {
    const result = await service.fetchListInfo({ postTime: Date.now() } as ExHentaiInfoItem);
    expect(result[0]).toMatchSnapshot();
  });

  it('getUrlFromPaginationInfo', async () => {
    const result = await service.getUrlFromPaginationInfo();
    expect(result).toEqual(mockDetail);
  });

  it('getAllThumbnaiUrls', async () => {
    const result = await service.getAllThumbnaiUrls();
    expect(result).toEqual(mockDetail);
  });

  it('fetchImageUrls', async () => {
    const result = await service.fetchImageUrls(['test1', 'test2']);
    expect(result).toEqual([mockComicNamePrefix + mockComicNameSuffix, mockComicNamePrefix + mockComicNameSuffix]);
  });

  it('downloadImages', async () => {
    await service.downloadImages(mockImgUrl, 'prefix');
    expect(ensureFileSync).toHaveBeenCalled();
    await service.downloadImages(mockImgUrl, 'prefix', true);
    expect(ensureFileSync).toHaveBeenCalled();
  });

  it('getThumbnailUrlFromDetailPage', async () => {
    const result = await service.getThumbnailUrlFromDetailPage(mockImgUrl);
    expect(result).toEqual([...mockDetail, ...mockDetail]);
  });
});
