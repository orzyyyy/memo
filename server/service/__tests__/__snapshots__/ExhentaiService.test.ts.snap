// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExhentaiService fetchListInfo 1`] = `
Object {
  "detailUrl": "https://exhentai.org/g/1476181/884c10c131/",
  "postTime": 1567369380000,
  "thumbnailUrl": "https://exhentai.org/t/60/de/60de2c9b39899a3fd7919dda44b6d7b6492cce1a-1853947-1720-2428-jpg_250.jpg",
}
`;
